File: default-implementation/observer-pattern.cs
using System;
using System.Collections.Generic;

namespace AnaliseDeSistemas
{
    interface IObserver
    {
        void DoSomething(string @event);
    }

    class SomeInterestingClass
    {
        private List<IObserver> observers = new List<IObserver>();

        public void AddObserver(IObserver o)
        {
            observers.Add(o);
        }

        public void NotifyObservers(string @event)
        {
            observers.ForEach(o => o.DoSomething(@event));
        }

        public void ActionA()
        {
            NotifyObservers("action A");
        }

        public void ActionB()
        {
            NotifyObservers("action B");
        }

        public void Work()
        {
            for (var i = 0; i < 10; i++)
            {
                Console.WriteLine("Im interesting and Im gonna do something...");

                if (new Random().Next(2) == 0)
                    ActionA();
                else
                    ActionB();

                Console.WriteLine();
            }
        }
    }

    class SomeCuriousClass : IObserver
    {
        private string name;

        public SomeCuriousClass(string name)
        {
            this.name = name;
        }

        public void DoSomething(string @event)
        {
            Console.WriteLine($"observer {@name} handled event {@event}");
        }
    }

    class Program
    {
        static void Main(string[] args)
        {
            var interesting = new SomeInterestingClass();

            interesting.AddObserver(new SomeCuriousClass("one"));
            interesting.AddObserver(new SomeCuriousClass("two"));
            interesting.AddObserver(new SomeCuriousClass("three"));

            interesting.Work();
        }
    }
}

File: default-implementation/observer-pattern.vb
Module AnaliseDeSistemas

	Public Interface IObserver
		Sub DoSomething(ByVal _event As String)
	End Interface

	Private Class SomeInterestingClass
		Private observers As New List(Of IObserver)

		Public Sub AddObserver(ByVal o As IObserver)
			observers.Add(o)
		End Sub

		Private Sub notifyObservers(ByVal _event As String)
			For Each o In observers
				o.DoSomething(_event)
			Next
		End Sub

		Public Sub ActionA()
			notifyObservers("action A")
		End Sub

		Public Sub ActionB()
			notifyObservers("action B")
		End Sub

		Public Sub Work()
			For i = 1 To 10
				Console.WriteLine("Im interesting and Im gonna do something...")

				Console.WriteLine(New Random().Next(2))

				If New Random().Next(2) = 0 Then
					ActionA()
				Else
					ActionB()
				End If

				Console.WriteLine()
			Next
		End Sub
	End Class

	Private Class SomeCuriousClass
		Implements IObserver

		Private name As String

		Public Sub New(ByVal name As String)
			Me.name = name
		End Sub

		Public Sub DoSomething(ByVal _event As String) Implements IObserver.DoSomething
			Console.WriteLine($"observer {name} handled event {_event}")
		End Sub
	End Class

	Sub Main()
		Dim interesting As New SomeInterestingClass

		interesting.AddObserver(New SomeCuriousClass("one"))
		interesting.AddObserver(New SomeCuriousClass("two"))
		interesting.AddObserver(New SomeCuriousClass("three"))

		interesting.Work()

		Console.ReadKey()
	End Sub

End Module
